Add tag_repo and upper-bound tools

timholy opened the pull request:
Tools:

- `tag_repo`: for creating/reintroducing an appropriate tag corresponding to a registered version of the package. (Needed when fixing ancient history.)
- `add_upperbounds`: add an upper bound for the most recent release
- `propagate_upperbounds`: propagate an upper bound to previous releases

tkelman commented on src/PkgDev.jl:
```diff
@@ -148,6 +161,55 @@ discover packages that may be due for tagging.
+"""
+    PkgDev.add_upperbound(pkg, dependent, version)
+
+Assuming package `pkg` dependents on package `dependent`, place an
```
say "`dependent` depends on `pkg`" ? dependent is not the clearest way to describe this

tkelman commented on src/PkgDev.jl:
```diff
@@ -148,6 +161,55 @@ discover packages that may be due for tagging.
+(as specified in the latest tagged version of `pkg`) backwards through
+previous tags. Any previous lower upper bounds are retained. At the
+conclusion of the process, all previous versions of `pkg` will have an
+upper bound for `dependent`.
```
will this be the case for previous versions that did not depend on dependent?

tkelman replied:
it often should be added if requirements have changed over time, but that's not always the right resolution

tkelman commented on src/entry.jl:
```diff
@@ -323,4 +344,77 @@ function freeable(io::IO = STDOUT)
+    i = find_dependent_index(lines, dependent, pkg)
+    vs = getvs(lines[i])
+    vs.upper == typemax(VersionNumber) || vs.upper == version ||
+        warn("replacing upper bound $(vs.upper) with $version")
```
this shouldn't be allowed to loosen existing bounds, only tighten

martinholters commented:
Should `tag_repo` backdate the tag to when the release was added to METADATA? Otherwise, tags added for ancient versions will suddenly appear on top of the Github Releases/Tags list, won't they? Not too terrible, but possibly confusing.

timholy commented:
Yes, that would be better. I'm not sure how to do that, though.

@tkelman, I've addressed your review comments locally, but also well aware that this needs tests. I'll get to it...

martinholters commented:
No idea how to do something like `git log -n 1 --format=%cD METADATA/${Package}/versions/${VERSION}/sha1` with libgit2. But with the date obtained, you'd have to pass a `Signature` with an accordingly set `time` to `tag_create`.

timholy commented:
OK, assuming this passes I think this is ready to go.

tkelman commented on src/PkgDev.jl:
```diff
@@ -153,6 +166,55 @@ discover packages that may be due for tagging.
+"""
+    PkgDev.add_upperbound(pkg, dependency, version)
+
+Assuming `pkg` depends` depends on `dependency`, place an upper bound
```
duplicate depends

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +377,80 @@ function freeable(io::IO = STDOUT)
+function find_dependency_index(lines, dependency, pkg)
+    idx = find(line->isa(line, Pkg.Reqs.Requirement) && line.package == dependency, lines)
+    isempty(idx) && error("package $pkg doesn't depend on $dependency")
+    length(idx) == 1 || error("package $pkg lists $dependency more than once")
```
this shouldn't be an error

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +377,80 @@ function freeable(io::IO = STDOUT)
+end
+
+function getvs(line)
+    length(line.versions.intervals) == 1 || error("expecting one version line, got $(length(line.versions))")
```
also not an error, can have multiple version intervals. should be able to intersect (0 - new upper) with the existing bounds

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +377,80 @@ function freeable(io::IO = STDOUT)
+function add_upperbound(pkg::AbstractString, dependency::AbstractString, version::VersionNumber)
+    # Parse REQUIRE and add the new upper bound
+    lines = cd(Pkg.dir(pkg)) do
+        Pkg.Reqs.read("REQUIRE")
```
this is pretty deeply flawed, it assumes the currently checked out package REQUIRE file matches the metadata requires file for the most recent tag.

timholy replied:
OK, please fix it properly.

tkelman replied:
best to treat the two files separately and add the bounds independently to each rather than overwriting one with the other, since they're independent

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +377,80 @@ function freeable(io::IO = STDOUT)
+        end
+        idx = find(line->isa(line, Pkg.Reqs.Requirement) && line.package == dependency, lines)
+        isempty(idx) && continue
+        length(idx) == 1 || error("package $pkg lists $dependency more than once")
```
not an error

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +382,78 @@ function freeable(io::IO = STDOUT)
+function find_dependency_index(lines, dependency, pkg)
+    idx = find(line->isa(line, Pkg.Reqs.Requirement) && line.package == dependency, lines)
+    isempty(idx) && error("package $pkg doesn't depend on $dependency")
+    idx[1]
```
why not return all of them and do `for i in find_dependency_index(lines, dependency, pkg)` each time this gets called?

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +382,78 @@ function freeable(io::IO = STDOUT)
+            Pkg.Reqs.read("requires")
+        end
+        idx = find(line->isa(line, Pkg.Reqs.Requirement) && line.package == dependency, lines)
+        isempty(idx) && continue
```
this is probably worth a warning, as old versions that didn't have the dependency that's being bounded could be downgraded to as a way of getting around the bound, and that may not always be expected or desired

tkelman commented on src/entry.jl:
```diff
@@ -354,4 +382,78 @@ function freeable(io::IO = STDOUT)
+                Pkg.Reqs.write("requires", lines)
+            end
+        else
+            upper = vs.upper
```
this seems odd - essentially if you encounter some old version that had a tighter upper bound than what you're trying to set, you propagate that tighter bound back even to older versions that didn't have it?
